### PR TITLE
Support for 128 bit bitwise operation

### PR DESCRIPTION
**Pull request by Sosutha**
* Added movh statement definition
* Added support for 128 bitwise operation
* Updated testcase and output

### CODE REVIEW

File: backends/dpdk/dpdkArch.h [hunk 1]
```diff
@@ -1101,10 +1101,23 @@ class ValidateOperandSize : public Inspector {
...
 
+    bool checkIf128bitOp(const IR::Expression *src1Op, const IR::Expression *src2Op) {
+        if (auto src1Type = src1Op->type->to<IR::Type_Bits>()) {
+            if (src1Type->width_bits() == 128) {
+                if (auto src2Type = src2Op->type->to<IR::Type_Bits>()) {
+                    if (src2Type->width_bits() == 128) return true;
+                }
+            }
+        }
+        return false;
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.h:1118]: Is it valid input to have operand1 of 128 bits and operand2 less than 128 bits at this stage?
If true, isValidOperandSize in line number 1106 may issue error for operand being greater than 64 bits.
IMO, isValidOperandSize function can be parameterized to check limit of 128 bits in exceptional cases.

File: backends/dpdk/dpdkHelpers.cpp [hunk 2]
```diff
@@ -1460,4 +1469,139 @@ bool ConvertStatementToDpdk::preorder(const IR::SwitchStatement *s) {
...
+    }
+}
+
+void ConvertStatementToDpdk::createSandboxHeader() {
+    auto fields = new IR::IndexedVector<IR::StructField>;
+    fields->push_back(new IR::StructField("upper_half", IR::Type_Bits::get(64)));
+    fields->push_back(new IR::StructField("lower_half", IR::Type_Bits::get(64)));
+    const IR::Type_Header *headerStruct = new IR::Type_Header(IR::ID("sandbox_header_t"), *fields);
+    auto name = new cstring("sandbox_header_t");
+    structure->header_types.emplace(*name, headerStruct);
```
Comment (psivanup) [inline backends/dpdk/dpdkHelpers.cpp:1596]:
```suggestion
    structure->header_types.emplace(cstring("sandbox_header_t"), headerStruct);
```

File: backends/dpdk/dpdkHelpers.cpp [hunk 3]
```diff
@@ -1460,4 +1469,139 @@ bool ConvertStatementToDpdk::preorder(const IR::SwitchStatement *s) {
...
+}
+
+void ConvertStatementToDpdk::createTmpVarForSandbox() {
+    auto fields = new IR::IndexedVector<IR::StructField>;
+    fields->push_back(new IR::StructField("tmp", IR::Type_Bits::get(64)));
+    const IR::Type_Header *headerStruct = new IR::Type_Header(IR::ID("tmp128_t"), *fields);
+    auto name = new cstring("tmp128_t");
+    structure->header_types.emplace(*name, headerStruct);
```
Comment (psivanup) [inline backends/dpdk/dpdkHelpers.cpp:1604]:
```suggestion
    structure->header_types.emplace(cstring("tmp128_t"), headerStruct);
```

**Review by psivanup — COMMENTED** [review]
Please check the comments

File: backends/dpdk/dpdkHelpers.cpp (hunk 3 above)
Comment (Sosutha) [reply backends/dpdk/dpdkHelpers.cpp]: Updated the code

**Comment by KrisNey-MSFT** [comment]
Suggest to try to compile DASH code as a test of completeness to see if the code compiles/builds?  If DASH has other than bitwise operations, it may fail.  Do not support additions to IPv6 addresses (we would likely not do this), but all else s/be supported.

File: backends/dpdk/dpdkArch.h (hunk 1 above)
Comment (ChrisDodd) [reply backends/dpdk/dpdkArch.h:1118]: Typechecking will ensure that the operands to any binary operation like these are the exact same type, so there's really no need to check both.  The only binops that can have different types for their operands are shifts, array index, and concat.

Comment (Sosutha) [reply backends/dpdk/dpdkArch.h]: Updated the code.

**Comment by jafingerhut** [comment]
I tried checking out this modified branch of the code, and built `p4c-dpdk`, then tried the following commands:

```
git clone https://github.com/sonic-net/DASH
cd DASH/dash-pipeline/bmv2
p4c-dpdk -DTARGET_BMV2_V1MODEL --target bmv2 --arch v1model dash_pipeline.p4
```

I see many warnings, which are unrelated to these changes, but I also see 6 errors "Unsupported bitwidth 128" as copied below:

```
stages/../dash_routing_types.p4(130): [--Werror=target-error] error: Unsupported bitwidth 128 in underlay_sip
                            underlay_sip = underlay_sip == 0 ? hdr.u0_ipv4.src_addr : (IPv4Addres...
                                           ^^^^^^^^^^^^
stages/../dash_routing_types.p4(130): [--Werror=target-error] error: Unsupported bitwidth 128 in 128w0
...                underlay_sip = underlay_sip == 0 ? hdr.u0_ipv4.src_addr : (IPv4Address)underla...
                                                  ^
stages/../dash_routing_types.p4(131): [--Werror=target-error] error: Unsupported bitwidth 128 in underlay_dip
                            underlay_dip = underlay_dip == 0 ? hdr.u0_ipv4.dst_addr : (IPv4Addres...
                                           ^^^^^^^^^^^^
stages/../dash_routing_types.p4(131): [--Werror=target-error] error: Unsupported bitwidth 128 in 128w0
...                underlay_dip = underlay_dip == 0 ? hdr.u0_ipv4.dst_addr : (IPv4Address)underla...
                                                  ^
stages/../dash_routing_types.p4(203): [--Werror=target-error] error: Unsupported bitwidth 128 in overlay_sip_mask
...( (IPv6Address)hdr.u0_ipv4.src_addr & ~overlay_sip_mask) | overlay_sip) & ~meta.eni_data.pl_si...
                                          ^^^^^^^^^^^^^^^^
[--Werror=target-error] error: Unsupported bitwidth 128 in meta._eni_data_pl_sip_mask20
dash_pipeline.p4(229): [--Werror=target-error] error: Unsupported bitwidth 128 in meta._overlay_data_dip_mask112
        routing_action_apply.apply(hdr, meta);
                                        ^^^^
dash_pipeline.p4(229): [--Werror=target-error] error: Unsupported bitwidth 128 in meta._overlay_data_sip_mask111
        routing_action_apply.apply(hdr, meta);
                                        ^^^^
```
Do you see these as well, Sosutha?

It would be good to understand why these errors occur, and if possible, either modify this PR so that those errors do not occur, or figure out a modified version of the DASH program that behaves as it is written today, but does not give those errors.

**Comment by Sosutha** [comment]
> [p4_16_samples/dash/dash-pipeline-pna-dpdk.p4](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/dash/dash-pipeline-pna-dpdk.p4)

Andy, I am getting same errors as you. This PR addresses only 128 bitwise operations. I can add implementation for comparison operations to this same PR.  So errors related to that will be solved with that implementation. For error in apply statement, I will have to check and revert back

**Comment by fruffy** [comment]
> > [p4_16_samples/dash/dash-pipeline-pna-dpdk.p4](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/dash/dash-pipeline-pna-dpdk.p4)
> 
> Andy, I am getting same errors as you. This PR addresses only 128 bitwise operations. I can add implementation for comparison operations to this same PR. So errors related to that will be solved with that implementation. For error in apply statement, I will have to check and revert back

We can also add the failing compilation to XFails for now. The important part is that this program is added as a test that is checked.

**Comment by Sosutha** [comment]
> > > [p4_16_samples/dash/dash-pipeline-pna-dpdk.p4](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/dash/dash-pipeline-pna-dpdk.p4)
> > 
> > 
> > Andy, I am getting same errors as you. This PR addresses only 128 bitwise operations. I can add implementation for comparison operations to this same PR. So errors related to that will be solved with that implementation. For error in apply statement, I will have to check and revert back
> 
> We can also add the failing compilation to XFails for now. The important part is that this program is added as a test that is checked.

@fruffy @jafingerhut Now, support for bitwise, comparison(==, !=) and compliment operations needed for DASH Program are added. Also failing compilation is added to Fails now. Now the errors are due to 128 bit constants in DASH P4 program which are not supported as of now. This PR can be merged  with these operations and handling 128 bit constants can be taken as separate PR/issue.

**Review by fruffy — APPROVED** [review]
Great, thank you adding the program as test. We can periodically update it. 
The MacOS failure can be fixed with a rebase.

**Comment by KrisNey-MSFT** [comment]
[like]  Kristina Moore reacted to your message:
________________________________

**Review by psivanup — APPROVED** [review]
Looks fine to me

**Pull request merged**